feat(cc-orga-member-*): init new components

florian-sanders-cc opened the pull request:
# Context

Fixes #376 
Fixes #411 
Fixes #412 
Fixes #629 
Fixes #630
Fixes #631 

# TODO

- [x] sync with @pdesoyres-cc to get the right `email.js` file.
- [x] remove demo-smart modifications.
- [x] remove demo-smart-${name} folders.
- [x] squash commits.

roberttran-cc commented:
Surprisingly, I like the dedicated section for leaving the organisation. :wink: 

About the supporting text, we can warn the user that the action is irreversible and they will have to ask permission to re-enter the organization.

> Where should we display this new area ? Looks better when at the very end but this means that it will be quite far down the page if the org contains a lot of members (fairly rare case though).

Agree with bottom.

> What should we do for the current user card when in admin mode ? Should we display a "leave" button within the card anyway ?

Leave button looks good.

> Should the current user be able to edit themself or not ?

I guess it depends on use cases (current ones and which one we want to enforce) and what the API authorizes.

> Depending on what we choose for the leave button when current user is admin / manager, we may end up with the current user card containing only an edit button or nothing at all ?!

Maybe we can use insights from the support team or sales on this one!

pdesoyres-cc commented:
> Where should we display this new area ?
I'm fine with positioning it at the bottom.

> What should we do for the current user card when in admin mode ?
Maybe a button in _link_ mode just after the 'Your account' badge?

> Should the current user be able to edit themself or not?
I would let the user edit himself

hsablonniere commented:
I really like this "leave the org" block over the single loneley delete button.

About the text:

* Can we try Danger zone in red like we do for the cc-addo-nadmin?
* Maybe something mentioning there is no confirmation and it's instant/immediate...
* The warning for admins is a bit too much but why not. I mean, if the user is the last admin, it will fail, right?
  * Do we display an error message in this block if you try to do this?

> Where should we display this new area ?

I prefer the block at the end. It's not that common to leave an org and I think we can expect people to scroll if they're looking for this.

> What should we do for the current user card when in admin mode ? Should we display a "leave" button within the card anyway ?

Keep the leave button.

> Should the current user be able to edit themself or not ? Depending on what we choose for the leave button when current user is admin / manager, we may end up with the current user card containing only an edit button or nothing at all ?!

* Non admin users should not be able to change their role
  * => no edit button
* Non admin users should be able to leave an org but with the bottom "leave org" block
  * => no delete button
* Admin users will have edit and delete buttons for each member
  * The label for the current user will be "leave" instead of "delete" and an admin can downgrade their role if they're not the last admin (as already implemented if I'm correct)

Galimede commented:
> Where should we display this new area?
Agree with bottom too.

> What should we do for the current user card when in admin mode? Should we display a "leave" button within the card anyway?

Leave button seems good.

> Should the current user be able to edit themself or not? Depending on what we choose for the leave button when current user is admin / manager, we may end up with the current user card containing only an edit button or nothing at all?!

I guess non admin users shouldn't have an edit button nor a delete button as they will be able to leave it at the bottom.
Admins should be able to edit and downgrade themselves, but it should be only effective if there is another admin, I guess. But if I remember well that's already the case with what you have done?

By the way in the UI how an admin can delete its organization? 🤔

Galimede commented:
> Where should we display this new area?
Agree with bottom too.

> What should we do for the current user card when in admin mode? Should we display a "leave" button within the card anyway?

Leave button seems good.

> Should the current user be able to edit themself or not? Depending on what we choose for the leave button when current user is admin / manager, we may end up with the current user card containing only an edit button or nothing at all?!

I guess non admin users shouldn't have an edit button nor a delete button as they will be able to leave the organization at the bottom.
However, admins should be able to edit and downgrade themselves, but it should be only effective if there is another admin, I guess. But if I remember well that's already the case with what you have done?

By the way, what about an admin that would want to delete the organization?

florian-sanders-cc commented:
Answers to questions about the "Danger zone" (discussed in sync)
> What do we do with the "Danger zone" when the list fails to load (error state) ?
Nothing, like for invite form. Users may click the button and trigger a toast with a generic error ("Something went wrong...").

> What do we do with the "Danger zone" when the list is loading (loading state) ?
Nothing, like for invite form. Users may click the button and leave the org before the list is even loaded or an loading error is displayed.

> What do we do with the component and its state when users leave the org ?
We fire a custom event to warn the console.
We set the state to "loaded" and remove the user from the org.

** Note **: 

- [ ] create a new "Event" section in the smart documentation. 
- [ ] update the smart doc to reflect this new "Event" section (may be used later on when we emit events to see what features are used, etc.).

florian-sanders-cc commented on src/mixins/with-resize-observer/with-resize-observer.js:
```diff
@@ -31,12 +31,19 @@ export function withResizeObserver (ParentClass) {
+        else {
+          window.requestAnimationFrame(() => {
+            this._onResize({ width });
+          });
```
Thanks @pdesoyres-cc for this :heart: 

This fixes a layout shift that we noticed in `cc-orga-member-list` :+1:

Galimede commented on src/components/cc-orga-member-card/cc-orga-member-card.types.d.ts:
```diff
@@ -0,0 +1,68 @@
+  | OrgaMemberCardStateUpdating
+  | OrgaMemberCardStateDeleting;
+
+interface OrgaMemberCardStateLoading {
```
Watch out  `OrgaMemberCardStateLoading` type is unused and might not be relevant anymore I guess? :thinking:

Galimede commented on src/components/cc-orga-member-card/cc-orga-member-card.js:
```diff
@@ -0,0 +1,531 @@
+  }
+
+  /* The accessible name provides more info than the visible text. It mentions the member being edited / to be edited. */
+  _getFirstBtnAccessibleName () {
```
When you first read the component code `_*firstBtn*` and `_*secondBtn*` functions might seem unclear as you don't know at first that it is related to the edit button and the leave one. Do we want to change it to something more clearer for the future? :thinking:

florian-sanders-cc replied:
These buttons used to have classes "first-btn" and "second-btn" but I didn't need them anymore and I removed the classes.
I agree with you that it's not very clear but I can't change it to "edit" for instance because this button can be used to edit and cancel.
The second button is even worse because it might be used for delete, leave or validate.

This problem will disappear if we use 4 buttons within the DOM to solve the icon issue for slow network. I'll wait until we make a decision on this subject and if we keep the current system with 2 buttons, I'll try to find something clearer for the names :+1:

Galimede commented on src/components/cc-orga-member-card/cc-orga-member-card.js:
```diff
@@ -0,0 +1,531 @@
+
+    /** @protected */
+    this.breakpoints = {
+      // used to switch to buttons wrapping below when component width < 740 and vertical card layout when < 580, buttons and badges below each other when < 350.
```
This line is quite hard to read could we replace `<` and `>` by below and above and make the sentence more readable?
I would suggest something like this but its up to you:
```suggestion
     /* used to: 
     * * wrap buttons when the component width is below 740.
     * * have a vertical card layout when the width is below 580.
     * * have buttons and badges below each other when the width is below 350.
     */
```

Galimede commented on src/components/cc-orga-member-card/cc-orga-member-card.js:
```diff
@@ -0,0 +1,531 @@
+    /** @type {Ref<CcSelect>} */
+    this._roleRef = createRef();
+
+    /** @type {null} Set by `withResizeObserver` mixin. See the `onResize` method for more info. */
```
Do you know why the type here is `null` when the variable is initialized as an empty string? :thinking: 
I would suggest to change it to `string` instead of `null` as resize observer will provide a string anyway.

Galimede commented on src/components/cc-orga-member-list/cc-orga-member-list.types.d.ts:
```diff
@@ -0,0 +1,52 @@
+  state: 'error';
+}
+
+interface InviteMember {
```
I have seen that InviteMember is used in the event for both of the components and you import DeleteMember here.
InviteMember is not present in the card types though.

With the current plugin and for the future I would suggest that you put both InviteMember and DeleteMember in the common type file and we will import them in the component directly when needed.

florian-sanders-cc replied:
`DeleteMember` is used by both indeed so I'll move it to common :+1: 
`InviteMember` is only used by the `cc-orga-member-list` so I'll leave it as it is if that's ok with you :+1:

Galimede commented on src/translations/translations.fr.js:
```diff
@@ -483,6 +483,69 @@ export const translations = {
+  'cc-orga-member-list.edit.error-unauthorised': ({ memberIdentity }) => sanitize`Vous n'êtes pas autorisé à modifier le membre <strong>${memberIdentity}</strong>.`,
+  'cc-orga-member-list.edit.success': ({ memberIdentity }) => sanitize`Le rôle de <strong>${memberIdentity}</strong> a bien été modifié.`,
+  'cc-orga-member-list.error': `Une erreur est survenue pendant le chargement de la liste des membres.`,
+  'cc-orga-member-list.filter.mfa': `Seulement 2FA désactivée`,
```
I don't know if I'm the only one but it sounds very weird in french :thinking:

florian-sanders-cc replied:
I agree, if we invert the words, does it sound better ? "2FA désactivée seulement" ?
I don't have other ideas right now :roll_eyes:

florian-sanders-cc replied:
I like it, a little bit longer but I find it very clear, I'll test it :+1:

Galimede reviewed (changes requested):
Aloha,

First of all @florian-sanders-cc, nice work and a big gg that's some impressive work and we all saw through this component that it was much more difficult that anticipaed so again a big gg! :hugs: :tada: 

As for my (short) review you will find everything in the github "posts". 

To keep track of it we currently have:
* a bug with the badges (role and 2FA), they are not aligned
* for the future with slow internet connection the icons of the actions button on the card take a time to switch when switching from editing mode to non editing mode.

roberttran-cc reviewed (changes requested):
:muscle: :muscle: :muscle: 
Finally! After much (needed) iterations, we're almost there! I only have a few feedback with some minor fixes.
GG for the work accomplished: this component has so much details, your implementation handles them so nicely. :tada:

florian-sanders-cc commented on src/components/cc-orga-member-card/cc-orga-member-card.js:
```diff
@@ -0,0 +1,531 @@
+ * This component heavily relies on `cc-stretch` to make sure all cards look the same whatever the role and MFA status may be.
+ * This component also heavily relies on CSS `grid` and the `withResizeObserver` mixin to switch from a "table" like design to a card design when the card width shrinks.
+ *
+ * @cssdisplay block
```
It will be block again because we're going to add a wrapper so I'll leave it like this but nice catch :smile:

florian-sanders-cc commented on src/components/cc-orga-member-card/cc-orga-member-card.types.d.ts:
```diff
@@ -0,0 +1,69 @@
+}
+/*endregion*/
+
+type LastAdminError = 'last-admin-leave' | 'last-admin-delete' | 'last-admin-update';
```
Indeed, but it won't be anymore with the changes ahead :joy: 
Nice catch again but I'll leave it there because I am going to need it :+1:

florian-sanders-cc replied:
Actually, I removed it cause I'm not exactly sure what it's gonna look like with the next changes but at least I'll start with something clean.
Thanks for this :+1:

florian-sanders-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,590 @@
+                outlined
+                @cc-button:click=${this._onLeave}
+                ?disabled="${this.members.lastAdminLeavingError}"
+                ?waiting="${this.members.state === 'leaving'}"
```
We used to, at the very beginning of this component but we cannot anymore because we use icon only buttons in desktop.
This is also why we decided in the future we would need to add modals :+1:

roberttran-cc replied:
Ho, makes sense! :+1:

florian-sanders-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.smart.md:
```diff
@@ -0,0 +1,62 @@
+
+| Name                              | Payload | Details                                                                                                       |
+|-----------------------------------|---------|---------------------------------------------------------------------------------------------------------------|
+| `cc-orga-member-list-smart:leave` |         | Fired when the current user leaves the organisation.<br/>Should be used to redirect the user to another page. |
```
This is confusing, even I thought I made a mistake at the beginning but this is related to events fired by the smart component.

Maybe I should make it clearer with the name but I don't really know how. Maybe I could ditch the name of the component in the name and use something like "orga-member:leave"?
@hsablonniere If you have ideas for this?

Also, should we rename and use this section to list events that the smart component listens to as well?

hsablonniere replied:
This is supposed to be a doc for users of the smart component so I wouldn't list the events listened by the smart component. I'm also thinking why not but if we want this, let's create an issue to discuss it.

We will probably remove the prefix of all of our events (we need to remove the `:` anyway) so I'm not sure yet.

I think this very event sent from smart could be prefix less and be `orga-member-leave-success`. I added a `-success` because it's fire only if the leave succeeded.

florian-sanders-cc replied:
Makes sense, thanks :+1:

florian-sanders-cc commented on src/translations/translations.fr.js:
```diff
@@ -483,6 +483,69 @@ export const translations = {
+  'cc-orga-member-card.mfa-enabled': `2FA activée`,
+  'cc-orga-member-card.role.accounting': `Comptable`,
+  'cc-orga-member-card.role.admin': `Admin`,
+  'cc-orga-member-card.role.developer': `Développeur`,
```
This is tricky but such a nice catch :+1: :clap: 

Let's try to find the best solution together in sync once I'm done process the rest of UI changes :+1:

florian-sanders-cc replied:
One of the interesting solutions that came up was to use words to described the activity more than the role: "development", etc.

We decided not to deal with this subject just yet since it's bigger than just this component.

florian-sanders-cc commented on src/translations/translations.en.js:
```diff
@@ -470,6 +470,69 @@ export const translations = {
+  'cc-orga-member-card.btn.validate.accessible-name': ({ memberIdentity }) => `Validate the edit of the member - ${memberIdentity}`,
+  'cc-orga-member-card.btn.validate.visible-text': `Validate`,
+  'cc-orga-member-card.current-user': `Your account`,
+  'cc-orga-member-card.error.last-admin': () => sanitize`<p><strong>This is the last admin of the organisation</strong></p><p>Please add a new admin before you can remove or edit this member.</p>`,
```
I'd say repetition is ok, let's keep the same word for the same thing, in technical English it's better, even for simple words :+1: 

Let's split right now indeed :+1:

pdesoyres-cc reviewed (changes requested):
This component is quite complex and you managed to handle it! Well done :clap:

florian-sanders-cc commented:
# Sync about error messages (last admin + unauthorised actions)

## Context

At this point, we had already decided to:

* show an error message (notice like) inside the card when the current user tries to edit or delete themselves from the org [Last admin scenari],
* show an error message (notice like) inside the card when the current user is manager and tries to delete / edit an admin [Unauthorised].

## Issue

When do we reset these error messages ?

## Possible solutions & discussions

### Prevent the user from doing something he cannot do anyway
  
#### disable the controls within their own card from the start

We try to only disable stuff after a user action so that users always know why controls are disabled so the first option is not great here.

#### hide the controls within their own card from the start

If we hide the controls, it means some cards will contain buttons, others won't. It is something we decided to avoid earlier when we introduced the authorisations (a simple user would have a leave button within their own card but other members wouldn't have any button inside it. This is why we introduced a "Danger Zone").

### Reset error messages everytime the list changes

  * everytime the list changes whatever the change may be ? (since we're using big objects for our state, it changes everytime anything happen in any of the card of the list)
  * everytime a delete / update has gone through (the smart component would reset the UI component).

Both would be acceptable but we're not sure about the layout shifts it would trigger.

### Use a toast for unauthorised actions

By using a toast for unauthorised actions, we don't have to worry about resetting the message.
It makes sense in this case because unauthorised actions are generally handled by the API and we usually use toasts for API responses.
It is consistent with what we do for invites (manager tries to invite an admin), they get an authorised action error toast.


### reset last admin error when new admin is added

The last admin error can be reset easily when the list changes and when it contains more than 1 admin.

## Conclusion

We decided to:

* use a toast for unauthorised actions
* reset the last admin error when a new admin is added.

pdesoyres-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.smart.js:
```diff
@@ -0,0 +1,252 @@
+     * @param {OrgaMemberRole} memberToUpdateRole - the current role of the member to update
+     * @return {boolean} - `true` if a manager is trying to edit and admin / `false` otherwise.
+     */
+    function checkIsManagerEditingAdmin (memberToUpdateRole) {
+      const currentUser = component.members.value.find((member) => member.isCurrentUser);
+      const isCurrentUserManager = currentUser.role === 'MANAGER';
+      const isMemberToUpdateAdmin = memberToUpdateRole === 'ADMIN';
+
+      return isCurrentUserManager && isMemberToUpdateAdmin;
+    }
```
you can:
- rename the function
- optimize it be failing faster when given role is not 'ADMIN'

```suggestion
    function isManagerEditingAdmin (role) {
      if (role !== 'ADMIN') {
        return false;
      }
      const currentUser = component.members.value.find((member) => member.isCurrentUser);
      return currentUser.role === 'MANAGER';
    }
```

pdesoyres-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.smart.js:
```diff
@@ -0,0 +1,252 @@
+            members.dangerZoneState = 'idle';
+          });
+          updateAuthorisations();
+          const leaveEvent = new Event('orga-member-leave-success');
+          window.dispatchEvent(leaveEvent);
```
```suggestion
          window.dispatchEvent(new Event('orga-member-leave-success'));
```

pdesoyres-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.smart.js:
```diff
@@ -0,0 +1,252 @@
+          if (error.id === UNAUTHORISED_ADMIN_ADDITION) {
+            notifyError(component, i18n('cc-orga-member-list.error.unauthorised.text'), i18n('cc-orga-member-list.error.unauthorised.heading'));
+          }
+          else if (error.message === 'You have performed that request too much.') {
```
move `'You have performed that request too much.'` into a constant near `UNAUTHORISED_ADMIN_ADDITION`

pdesoyres-cc reviewed (changes requested):
@florian-sanders-cc , you've done really hard work, well done! :muscle:

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+   * Check if the member is the last admin of the organisation.
+   *
+   * @param {OrgaMemberState} member - the member to check
+   * @return {boolean} - true if the current user is an admin and there is only one admin left in the organisation
```
I prefer `current` over `given` but maybe it's just me.

florian-sanders-cc replied:
To talk about the connected user I prefer `current` but what you are both pointing out is that this method is in a weird state following my last modifications.
It can be used to check that any member is not the last admin, but it still mentions "current user" in its body / JSDoc.

I'll make it more generic so that it is consistent.

It may be used to check any member so I'll go for "given member" and replace `isCurrentUserAdmin` by `isMemberAdmin` within the method.

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+  }
+
+  /**
+   * Check if the member is the last admin of the organisation.
```
I prefer `current` over `given` but maybe it's just me.

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+   * Set or remove an error of the member card corresponding to the provided member id.
+   *
+   * @param {string} memberId - the id corresponding to the member card
+   * @param {boolean} errorType - whether the card should display an error message or not
```
Also, it seems like you're only using this `_setMemberCardError` twice and the second param is always `true`.

florian-sanders-cc replied:
Yes, since we no longer show errors within the card when managers try to temper with an admin, this method doesn't make much sense.

I removed it and I set the errors directly from `_onUpdateFromCard` and `_onLeave...` methods.

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+
+  _onInviteMember () {
+    const existingEmails = this.members.value.map((member) => member.email);
+    const email = this._inviteMemberEmailRef.value.value;
```
We should and we should apply the same as the emails panel.

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+ * @cssdisplay block
+ *
+ * @event {CustomEvent<InviteMember>} cc-orga-member-list:invite - Fires the `email` and `role` information inside an object whenever the invite button is clicked.
+ * @event {CustomEvent<DeleteMember>} cc-orga-member-list:leave - Fires the `id` of the member to be removed from the org with their `name` or `email`.
```
> Since Admin is already a simplification of Administrator, could we use "Dev" ?

I'm not a great fan of this :disappointed:

hsablonniere commented on src/components/cc-orga-member-card/cc-orga-member-card.stories.js:
```diff
@@ -0,0 +1,242 @@
+  isCurrentUser: false,
+};
+
+const longEmail = 'very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-email-address@very-very-very-very-very-very-very-long-example.com';
```
Tiny detail but please use a subdomain of `example.com`.

=> `very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-very-long-email-address@very-very-very-very-very-very-very-long.example.com`

hsablonniere commented on src/components/cc-orga-member-list/cc-orga-member-list.smart.md:
```diff
@@ -0,0 +1,62 @@
+| Method   | Type                                 | Cache?  |
+|----------|:-------------------------------------|---------|
+| `GET`    | `/v2/self/id`                        | Default |
+| `GET`    | `/v2/organisations/${orgId}/members` | Default |
+| `PUT`    | `/v2/organisations/${orgId}/members` | Default |
+| `DELETE` | `/v2/organisations/${orgId}/members` | Default |
```
In other documents we use `id` inside curly braces without the `$`:

```
| `GET`    | `/v2/organisations/{id}/members` | Default |
| `PUT`    | `/v2/organisations/{id}/members` | Default |
| `DELETE` | `/v2/organisations/{id}/members` | Default |
```

hsablonniere reviewed (changes requested):
Impressive work !!

* :+1: Wow the resize-observer thing is tricky :p
* You have several commits where the `Fixes #XYZ` is not in the body
* You will remove the tmp demo smart folders, right?
* list: I know it's a detail but can we order the sub renders in the same order they're used in the main render ?

florian-sanders-cc commented on src/components/cc-orga-member-list/cc-orga-member-list.js:
```diff
@@ -0,0 +1,632 @@
+      return a.email.localeCompare(b.email, { sensitivity: 'base' });
+    });
+
+    return sortedFilteredMemberList;
```
I found chaining of filter + sort a bit difficult to read because they both contain a few lines of code so I decided to keep the two steps like this, hope it's ok :wink:

hsablonniere commented on src/translations/translations.fr.js:
```diff
@@ -515,6 +515,71 @@ export const translations = {
+  'cc-orga-member-list.leave.btn': `Quitter l'organisation`,
+  'cc-orga-member-list.leave.error': `Une erreur est survenue lorsque vous avez tenté de quitter l'organisation.`,
+  'cc-orga-member-list.leave.error-last-admin.heading': `Vous êtes le dernier admin de l'organisation`,
+  'cc-orga-member-list.leave.error-last-admin.text': `Veuillez designer un nouvel admin avant de pouvoir quitter l'organisation.`,
```
Misses an accent on "désigner"

```suggestion
  'cc-orga-member-list.leave.error-last-admin.text': `Veuillez désigner un nouvel admin avant de pouvoir quitter l'organisation.`,
```
